src/test/java/ibm/jceplus/junit/base/BaseTestAESCopySafe.java

```diff
@@ -0,0 +1,182 @@
+        Arrays.fill(clearText, 0, INPUT_LENGTH, (byte)0x01);
+        System.arraycopy(clearText, 0, workingBuffer, 0, INPUT_LENGTH);
+
+        // Get baseline encrypted value. This baseline will be used through the rest of the 
+        // test to be a known answer for cipher text.
+        initCipher(c, mode, true);
+        if (DEBUG) {
+            System.out.println("Calling c." + (isUpdate ? "update" : "doFinal") + "(workingBuffer, 0, INPUT_LENGTH)");
+            System.out.println("    INPUT_LENGTH: " + INPUT_LENGTH);
+            System.out.println("    workingBuffer:\n" + BaseUtils.bytesToHex(workingBuffer));
+        }
+        byte[] cipherText = null;
+        if (isUpdate) {
+            cipherText = c.update(workingBuffer, 0, INPUT_LENGTH);
+        } else {
+            cipherText = c.doFinal(workingBuffer, 0, INPUT_LENGTH);
+        }
+        if (DEBUG) {
+            System.out.println("cipherText:\n" + BaseUtils.bytesToHex(cipherText));
+            System.out.println ("cipherText.length: " + cipherText.length);
+        }
+
+        // A GCM cipher must be reinitialized since the IV is typically not supposed to be reused. This
+        // test however expects a deterministic encrypted value in order to evaluate if the operation
+        // worked as expected.
+        initCipher(c, mode, true);
+
+        System.arraycopy(clearText, 0, workingBuffer, inputOffset, INPUT_LENGTH);
+        if (DEBUG) {
+            System.out.println ("Calling c." + (isUpdate ? "update" : "doFinal") + "(workingBuffer, inputOffset, INPUT_LENGTH, workingBuffer, outputOffset)");
+            System.out.println("    inputOffset:" + inputOffset);
+            System.out.println("    INPUT_LENGTH:" + INPUT_LENGTH);
+            System.out.println("    outputOffset:" + outputOffset);
+            System.out.println("    workingBuffer:\n" + BaseUtils.bytesToHex(workingBuffer));
+        }
+        if (isUpdate) {
+            c.update(workingBuffer, inputOffset, INPUT_LENGTH, workingBuffer, outputOffset);
+        } else {
+            c.doFinal(workingBuffer, inputOffset, INPUT_LENGTH, workingBuffer, outputOffset);
+        }
+        if (DEBUG) {
+            System.out.println("workingBuffer:\n" + BaseUtils.bytesToHex(workingBuffer));
+        }
+        
+        assertArrayEquals(Arrays.copyOfRange(workingBuffer, outputOffset, outputOffset + cipherText.length), cipherText, "Encryption check failed.");
+        if (DEBUG) {
+            System.out.println("Encrypt check passed.");
+        }
+
+        // Test decryption now, we should get back the original clear text value as a result.
+        if (DEBUG) {
+            System.out.println("c.init(DECRYPT) with KEY and Params");
+        }
+        initCipher(c, mode, false);
+
+        Arrays.fill(workingBuffer, 0, INPUT_LENGTH, (byte)0x00);
+        System.arraycopy(cipherText, 0, workingBuffer, inputOffset, cipherText.length);
+
+        if (DEBUG) {
```

Comment:
Should we specify that we are testing decryption in this step?

Reply:
Yes its prob best to state we are doing encryption or decryption in this DEBUG statement.  Adjusted this in new commit https://github.com/IBM/OpenJCEPlus/pull/295/commits/46a71f1f5e9d53cc94e3336b27fbf231deaeeab3 .